File: go/algorithms/radix_lsd.go
package algorithms

// RadixSortLSD sorts in-place using least-significant-digit radix sort; supports negatives via offset.
func RadixSortLSD(arr []int) {
    if len(arr) == 0 {
        return
    }
    min := arr[0]
    max := arr[0]
    for _, v := range arr {
        if v < min {
            min = v
        }
        if v > max {
            max = v
        }
    }
    if min < 0 {
        for i := range arr {
            arr[i] -= min
        }
        max -= min
    }
    for exp := 1; max/exp > 0; exp *= 10 {
        countingSortExp(arr, exp)
    }
    if min < 0 {
        for i := range arr {
            arr[i] += min
        }
    }
}

func countingSortExp(arr []int, exp int) {
    n := len(arr)
    output := make([]int, n)
    count := make([]int, 10)
    for i := 0; i < n; i++ {
        digit := (arr[i] / exp) % 10
        count[digit]++
    }
    for i := 1; i < 10; i++ {
        count[i] += count[i-1]
    }
    for i := n - 1; i >= 0; i-- {
        digit := (arr[i] / exp) % 10
        count[digit]--
        output[count[digit]] = arr[i]
    }
    copy(arr, output)
}

File: rust/src/radix_lsd.rs
pub fn radix_sort_lsd(arr: &mut [i32]) {
    if arr.is_empty() {
        return;
    }
    let mut min = arr[0];
    let mut max = arr[0];
    for &v in arr.iter() {
        if v < min {
            min = v;
        }
        if v > max {
            max = v;
        }
    }
    if min < 0 {
        for v in arr.iter_mut() {
            *v -= min;
        }
        max -= min;
    }
    let mut exp = 1;
    while max / exp > 0 {
        counting_sort_exp(arr, exp);
        exp *= 10;
    }
    if min < 0 {
        for v in arr.iter_mut() {
            *v += min;
        }
    }
}

fn counting_sort_exp(arr: &mut [i32], exp: i32) {
    let n = arr.len();
    let mut output = vec![0i32; n];
    let mut count = [0usize; 10];
    for &v in arr.iter() {
        let digit = ((v / exp) % 10) as usize;
        count[digit] += 1;
    }
    for i in 1..10 {
        count[i] += count[i - 1];
    }
    for &v in arr.iter().rev() {
        let digit = ((v / exp) % 10) as usize;
        count[digit] -= 1;
        output[count[digit]] = v;
    }
    arr.copy_from_slice(&output);
}
